xitorch/_impls/linalg/solve.py

```diff
@@ -323,6 +323,91 @@ def bicgstab(A: LinearOperator, B: torch.Tensor,
         xk = xk.transpose(0, -1).squeeze(0)  # (*, nr, ncols)
     return xk
 
+def gmres(A: LinearOperator, B: torch.Tensor,
+          E: Optional[torch.Tensor] = None,
+          M: Optional[LinearOperator] = None,
+          posdef: Optional[bool] = None,
+          max_niter: Optional[int] = None,
+          rtol: float = 1e-6,
+          atol: float = 1e-8,
+          eps: float = 1e-12,
+          **unused) -> torch.Tensor:
+    r"""
+    Solve the linear equations using Generalised minial residual method.
+
+    Keyword arguments
+    -----------------
+    posdef: bool or None
+        Indicating if the operation :math:`\mathbf{AX-MXE}` a positive
+        definite for all columns and batches.
+        If None, it will be determined by power iterations.
+    max_niter: int or None
+        Maximum number of iteration. If None, it is set to ``int(1.5 * A.shape[-1])``
+    rtol: float
+        Relative tolerance for stopping condition w.r.t. norm of B
+    atol: float
+        Absolute tolerance for stopping condition w.r.t. norm of B
+    eps: float
+        Substitute the absolute zero in the algorithm's denominator with this
+        value to avoid nan.
+    """ 
+    nr, ncols = A.shape[-1], B.shape[-1]
+    if max_niter is None:
+        max_niter = int(1.5 * nr)
+
+    # if B is all zeros, then return zeros
+    batchdims = _get_batchdims(A, B, E, M)
+    if torch.allclose(B, B * 0, rtol=rtol, atol=atol):
+        x0 = torch.zeros((*batchdims, nr, ncols), dtype=A.dtype, device=A.device)
+        return x0
+
+    # setup the preconditioning and the matrix problem
+    need_hermit = False
+    A_fcn, AT_fcn, B2, col_swapped = _setup_linear_problem(A, B, E, M, batchdims,
+                                                           posdef, need_hermit)
+    # get the stopping matrix
+    B_norm = B2.norm(dim=-2, keepdim=True)  # (*BB, 1, nc)
+    stop_matrix = torch.max(rtol * B_norm, atol * torch.ones_like(B_norm))  # (*BB, 1, nc)

```

Comment:
Where do you use `stop_matrix` in determining the stop condition? It should be used to respect user's rtol and atol inputs. You can see other implementation on how they use `stop_matrix` for the stopping conditions.

Reply:
fixed in 9b5a9ae6031e3e487deb6d87fc5e565f74af90d1